Atualizando o readme com as informações para o RA3

breno-rossi opened the pull request.

breno-rossi committed:
Atualizando o readme com as informações para o RA3

fbleyruthes reviewed (changes requested):
# Checklist de Melhorias: README.md - RA3

**Data:** 2025-11-01
**Status Atual:** 88% de conformidade com a rubrica
**Objetivo:** Atingir 100% de conformidade com a rubrica da Fase 3 (RA3)

---

## 🚨 CRÍTICO - Deve ser corrigido imediatamente

### 1. Arquivos de Saída Markdown Obrigatórios (NÃO DOCUMENTADOS)

**Rubrica - Seção 18.5 e 18.11.4:**
> "O repositório deve conter:
> - Arquivo markdown com a gramática de atributos
> - Arquivo markdown com árvore sintática da última execução do código
> - Arquivo markdown com os erros semânticos detectados
> - Arquivo markdown com o julgamento de tipos descrevendo as regras de dedução criadas e onde foram aplicadas na última execução do código"

**Problema:**
O README menciona apenas 1 dos 4 arquivos obrigatórios. Além disso, os arquivos estão divididos entre dois diretórios:
- `outputs/RA3/` - tem apenas 2 arquivos
- `relatorios/` - tem 5 arquivos markdown (incluindo os obrigatórios)

**Arquivos reais encontrados:**
- ✅ `outputs/RA3/gramatica_atributos.md` - Gramática (mencionado)
- ✅ `outputs/RA3/arvore_atribuida.json` - JSON (mencionado)
- ❌ `relatorios/arvore_atribuida.md` - NÃO mencionado (deveria estar em outputs/RA3/)
- ❌ `relatorios/erros_sematicos.md` - NÃO mencionado (deveria estar em outputs/RA3/)
- ❌ `relatorios/julgamento_tipos.md` - NÃO mencionado (deveria estar em outputs/RA3/)
- ❌ `relatorios/tabela_simbolos.md` - NÃO mencionado (extra, mas útil)

**Ação:**
Adicionar após **linha 83** (seção "Arquivos de Saída") uma nova subseção documentando os 4 arquivos obrigatórios:

**Título da seção:** Arquivos de Saída do Analisador Semântico (RA3)

**Conteúdo:**

O analisador semântico gera os seguintes arquivos markdown:

**Em outputs/RA3/:**
- `gramatica_atributos.md` - Gramática de atributos completa
- `arvore_atribuida.json` - Árvore sintática em JSON (interoperabilidade)

**Em relatorios/:**
- `arvore_atribuida.md` - Árvore sintática atribuída da última execução
- `erros_sematicos.md` - Relatório de erros semânticos detectados
- `julgamento_tipos.md` - Relatório de julgamento de tipos com regras aplicadas
- `tabela_simbolos.md` - Estado da tabela de símbolos
- `gramatica_atributos.md` - Cópia da gramática

**Nota:** Considerar consolidar todos os arquivos em `outputs/RA3/` para melhor organização.

**Impacto:** Perda de até 30% da nota se não documentado.

---

## ⚠️ IMPORTANTE - Deve ser adicionado

### 2. Estrutura do Projeto - outputs/RA3/ Incompleto

**Rubrica - Seção 18.11.4:**
> "O repositório deve conter arquivo markdown com a gramática de atributos, árvore sintática, erros semânticos e julgamento de tipos"

**Problema:**
A árvore de diretórios (linhas 184-185) não reflete a estrutura real. Os arquivos estão divididos entre `outputs/RA3/` e `relatorios/`.

**Ação:**
Atualizar **linhas 184-185** para mostrar estrutura REAL atual:

```
├── outputs/                            # Arquivos de saída gerados
│   └── RA3/                           # Saídas do analisador semântico
│       ├── gramatica_atributos.md          # Gramática de atributos
│       └── arvore_atribuida.json           # Árvore em JSON
│
└── relatorios/                         # Relatórios do analisador semântico
    ├── arvore_atribuida.md             # Árvore sintática atribuída
    ├── erros_sematicos.md              # Erros semânticos (nota: erro ortográfico)
    ├── julgamento_tipos.md             # Julgamento de tipos
    ├── tabela_simbolos.md              # Tabela de símbolos
    └── gramatica_atributos.md          # Gramática (duplicada)
```

**Recomendação adicional:** Consolidar todos em `outputs/RA3/` para conformidade com a rubrica.

---

### 3. Formato JSON da Árvore Atribuída

**Rubrica - Nota na Seção 18.7:**
> "A árvore sintática abstrata atribuída, além de estar na documentação, deve ser salva em JSON para facilitar a interoperabilidade com a próxima fase. O formato deve incluir pelo menos: tipo do vértice, tipo inferido, filhos, e número da linha."

**Problema:**
O README não especifica o formato JSON da árvore atribuída.

**Ação:**
Adicionar após seção "Arquivos de Saída" (linha 83) uma nova subseção:

**Título:** Formato da Árvore Sintática Atribuída (JSON)

**Conteúdo:**

A árvore é salva em JSON com os seguintes campos obrigatórios:

- `tipo_vertice` - Tipo do nó (EXPRESSAO_ARITMETICA, LITERAL, OPERADOR, etc.)
- `tipo_inferido` - Tipo semântico inferido (int, real, boolean)
- `numero_linha` - Linha do código fonte
- `filhos` - Lista de nós filhos (recursivo)

**Exemplo de estrutura JSON:**

```json
{
  "tipo_vertice": "EXPRESSAO_ARITMETICA",
  "tipo_inferido": "int",
  "numero_linha": 1,
  "filhos": [
    {"tipo_vertice": "LITERAL", "tipo_inferido": "int", "valor": 5},
    {"tipo_vertice": "LITERAL", "tipo_inferido": "int", "valor": 3},
    {"tipo_vertice": "OPERADOR", "operador": "+"}
  ]
}
```

Este formato facilita interoperabilidade com a próxima fase.

---

## 💡 INFORMATIVO - Recomendado adicionar

### 4. Testes Adicionais para Robustez

**Problema:**
O README documenta 3 arquivos de teste principais, mas não menciona os testes adicionais existentes.
Além disso, precisamos mencionar que utilizamos a biblioteca pytests para realizar testes unitários.

**Observação:**
Existem 11 arquivos de teste em `inputs/RA3/`:
- 3 principais já documentados (teste1_valido, teste2_erros_tipos, teste3_erros_memoria)
- 8 adicionais para testes de robustez e casos específicos

--> Mencionar também arquivos de testes do pytests!

**Ação:**
Adicionar nota após **linha 563** (final da seção "Arquivos de Teste"):

**Conteúdo sugerido:**

**Nota sobre testes adicionais:**
Além dos 3 arquivos de teste principais descritos acima, o projeto inclui 8 arquivos de teste adicionais para validar casos específicos e garantir robustez do analisador semântico (teste4_analisador_completo.txt, teste5_analisador_corrigido.txt, teste_res_completo.txt, teste_parser_elaborado.txt, teste_erro_tipo.txt, teste_debug_subtracao.txt, teste_debug_res.txt, teste_analise_hist.txt). Todos os testes podem ser executados com o mesmo comando `python3 compilar.py inputs/RA3/<arquivo>`.

---

### 5. Exemplos de Erros Semânticos Expandidos

**Problema:**
O README mostra apenas 1 exemplo de erro semântico (linha 610-612).

**Ação:**
Expandir seção "Tratamento de Erros > Erros Semânticos (RA3)" (linha 604-613) com mais exemplos de cada categoria.

**Substituir conteúdo atual por:**

### Erros Semânticos (RA3)

**Erros de tipos:**

Operações aritméticas com tipos incompatíveis:
```
ERRO SEMÂNTICO [Linha 5]: Operador '/' requer ambos operandos do tipo 'int', mas recebeu 'real' e 'int'
Contexto: (5.5 2 /)

ERRO SEMÂNTICO [Linha 8]: Operador '%' requer ambos operandos do tipo 'int', mas recebeu 'int' e 'real'
Contexto: (10 3.5 %)

ERRO SEMÂNTICO [Linha 12]: Operador '^' requer expoente do tipo 'int', mas recebeu 'real'
Contexto: (2 3.5 ^)
```

**Erros de memória:**

Variáveis não inicializadas ou tipos incompatíveis:
```
ERRO SEMÂNTICO [Linha 3]: Variável 'CONTADOR' utilizada sem inicialização
Contexto: (CONTADOR)

ERRO SEMÂNTICO [Linha 7]: Tipo 'boolean' não pode ser armazenado em memória
Contexto: ((5 3 >) X)
```

**Erros de controle:**

Estruturas de controle malformadas:
```
ERRO SEMÂNTICO [Linha 10]: IFELSE requer ambos os ramos com o mesmo tipo, mas recebeu 'int' e 'real'
Contexto: ((X 5 >) (10) (5.5) IFELSE)

ERRO SEMÂNTICO [Linha 15]: FOR requer inicio, fim e passo do tipo 'int'
Contexto: ((1.5) (10) (1) (corpo) FOR)
```

---

### 6. Nomes Exatos das Funções (Corrigir Inconsistências)

**Problema:**
O README lista nomes de funções que não correspondem exatamente às funções reais no código (linhas 570-588).

**Funções REAIS encontradas no código:**

**Aluno 1 (Gramática e Tabela de Símbolos):**
- `definirGramaticaAtributos()` ✅ Correto no README
- `inicializarTabelaSimbolos()` ❌ README lista `criar_tabela_simbolos()`
- Não há funções `adicionar_simbolo()` ou `buscar_simbolo()` standalone

**Aluno 2 (Análise de Tipos):**
- `analisarSemantica(arvoreSintatica, gramatica, tabela)` ✅ Correto
- `promover_tipo(tipo1, tipo2)` ✅ Existe
- Funções auxiliares: `tipos_compativeis_aritmetica()`, `tipos_compativeis_divisao_inteira()`, `tipo_resultado_aritmetica()`, etc.
- Não há função `verificar_tipos()` standalone

**Aluno 3 (Memória e Controle):**
- `analisarSemanticaMemoria(arvore_anotada, seqs_map, tabela)` ⚠️ Assinatura diferente no README
- `analisarSemanticaControle(arvore_anotada, seqs_map, tabela)` ⚠️ Assinatura diferente no README
- Não há função `verificar_inicializacao()` standalone

**Aluno 4 (Geração de Árvore e Relatórios):**
- `gerarArvoreAtribuida(arvoreAnotada)` ✅ Correto
- `salvarArvoreAtribuida(arvoreAtribuida)` ❌ NÃO listada no README
- `gerarRelatoriosMarkdown(arvoreAtribuida, errosSemanticos, ...)` ❌ NÃO listada no README
- Não há função `gerar_relatorios()` genérica

**Ação:**
Substituir seção "Funções Principais > RA3" (linhas 570-588) pelos nomes e assinaturas EXATOS:

### RA3 - Análise Semântica

**Gramática e Tabela de Símbolos (Aluno 1):**
- `definirGramaticaAtributos()` - Define gramática de atributos completa
- `inicializarTabelaSimbolos()` - Cria estrutura de tabela de símbolos
- `promover_tipo(tipo1, tipo2)` - Realiza promoção de tipos quando necessário

**Análise de Tipos (Aluno 2):**
- `analisarSemantica(arvoreSintatica, gramatica, tabela)` - Função principal de análise semântica
- `tipos_compativeis_aritmetica(tipo1, tipo2)` - Verifica compatibilidade aritmética
- `tipos_compativeis_divisao_inteira(tipo1, tipo2)` - Verifica divisão inteira
- `tipo_resultado_aritmetica(tipo1, tipo2, operador)` - Calcula tipo resultante

**Memória e Controle (Aluno 3):**
- `analisarSemanticaMemoria(arvore_anotada, seqs_map, tabela)` - Valida uso de memórias
- `analisarSemanticaControle(arvore_anotada, seqs_map, tabela)` - Valida estruturas de controle

**Geração de Árvore e Relatórios (Aluno 4):**
- `gerarArvoreAtribuida(arvoreAnotada)` - Gera árvore sintática atribuída
- `salvarArvoreAtribuida(arvoreAtribuida)` - Salva árvore em JSON
- `gerarRelatoriosMarkdown(arvoreAtribuida, errosSemanticos, ...)` - Gera todos os relatórios markdown

---

### 7. Precisão Numérica (Referência da Rubrica)

**Rubrica - Seção 18.2.1:**
> "A precisão dos números de ponto flutuante depende da arquitetura do processador onde o programa será executado para avaliação:
>
> - Se a arquitetura for de 8 bits: use meia precisão (16 bits, IEEE 754);
> - Se a arquitetura for de 16 bits: use precisão simples (32 bits, IEEE 754);
> - Se a arquitetura for de 32 bits: use precisão dupla (64 bits, IEEE 754);
> - Se a arquitetura for de 64 bits: use precisão quádrupla (128 bits, IEEE 754).
>
> As operações de divisão inteira e resto são realizadas exclusivamente com números inteiros. Além disso, lembre-se, para esta avaliação você deve considerar apenas a arquitetura do Arduino Uno R3, Arduino Mega, ou qualquer outro processador de 8 bits."

**Problema:**
O README não menciona precisão numérica e considerações de arquitetura.

**Ação:**
Adicionar após **linha 262** (seção "Tipos de Dados") uma nova subseção:

**Título:** Precisão Numérica

**Conteúdo:**

A precisão dos números de ponto flutuante depende da arquitetura do processador:

- **8 bits** (Arduino Uno R3/Mega): meia precisão (16 bits, IEEE 754)
- **16 bits**: precisão simples (32 bits, IEEE 754)
- **32 bits**: precisão dupla (64 bits, IEEE 754)
- **64 bits**: precisão quádrupla (128 bits, IEEE 754)

**Para esta avaliação:** Considerar arquitetura de **8 bits** (Arduino Uno R3, Arduino Mega).

**Observação:** As operações de divisão inteira (/) e resto (%) são realizadas exclusivamente com números inteiros.

---

### 8. Exemplo de Saída do Programa

**Rubrica - Seção 18.11.3:**
> "Exemplos de uso do programa"

**Problema:**
O README tem exemplos de entrada, mas não de saída do programa.

**Ação:**
Adicionar após **linha 93** (seção "Depuração") uma nova subseção:

**Título:** Exemplo de Saída

**Conteúdo:**

**Execução válida:**

```
$ python3 compilar.py inputs/RA3/teste1_valido.txt

=== ANÁLISE LÉXICA ===
✓ Tokens gerados: 150 tokens

=== ANÁLISE SINTÁTICA ===
✓ Árvore sintática gerada

=== ANÁLISE SEMÂNTICA ===
✓ Verificação de tipos: OK
✓ Verificação de memória: OK
✓ Verificação de estruturas de controle: OK

Arquivos gerados:
- outputs/RA3/gramatica_atributos.md
- outputs/RA3/arvore_atribuida.json
- relatorios/julgamento_tipos.md
- relatorios/arvore_atribuida.md

✓ Compilação concluída sem erros
```

**Execução com erros:**

```
$ python3 compilar.py inputs/RA3/teste2_erros_tipos.txt

=== ANÁLISE SEMÂNTICA ===
ERRO SEMÂNTICO [Linha 5]: Operador '/' requer ambos operandos do tipo 'int'
Contexto: (5.5 2 /)

✗ Compilação falhou com 1 erro semântico
Saída: relatorios/erros_sematicos.md
```

---

## ✅ CHECKLIST RESUMIDO

### Prioridade CRÍTICA (Obrigatório)
- [ ] **Item 1:** Documentar 4 arquivos markdown + estrutura real (outputs/RA3 + relatorios/) → Linha 83

### Prioridade IMPORTANTE (Fortemente Recomendado)
- [ ] **Item 2:** Atualizar estrutura do projeto mostrando outputs/RA3/ E relatorios/ → Linhas 184-185
- [ ] **Item 3:** Especificar formato JSON da árvore → Linha 83

### Prioridade INFORMATIVA (Recomendado)
- [ ] **Item 4:** Adicionar nota sobre testes adicionais de robustez → Linha 563
- [ ] **Item 5:** Adicionar mais exemplos de erros semânticos → Linha 604
- [ ] **Item 6:** Corrigir nomes exatos das funções (usar nomes reais do código) → Linhas 570-588
- [ ] **Item 7:** Adicionar precisão numérica (Arduino 8 bits) → Linha 262
- [ ] **Item 8:** Adicionar exemplo de saída → Linha 93

---

## 📊 Estimativa de Impacto

| Prioridade | Itens | Impacto na Nota | Tempo Estimado |
|------------|-------|-----------------|----------------|
| CRÍTICA | 1 | Até -30% | 20 min |
| IMPORTANTE | 2-3 | Até -10% | 15 min |
| INFORMATIVA | 4-8 | Até -10% | 25 min |
| **TOTAL** | **8 itens** | **Até -50%** | **60 min** |

---

## 📝 Notas Finais

1. **Conformidade atual:** 88% (rotinas de teste já documentadas ✅)
2. **Conformidade após correções:** 100%
3. **Referência completa:** `docs/RA3/documents/Comandos_Fase_3.pdf`
4. **README original:** Muito bem escrito, apenas faltam detalhes específicos da rubrica

**Descobertas importantes:**
- Arquivos markdown estão divididos entre `outputs/RA3/` (2 arquivos) e `relatorios/` (5 arquivos)
- Existem 11 arquivos de teste total (3 principais + 8 adicionais para robustez)
- Nomes de funções no README não correspondem exatamente aos nomes reais no código
- Arquivo `relatorios/erros_sematicos.md` tem erro de ortografia (deveria ser "semanticos")

breno-rossi committed:
Merge branch 'main' of https://github.com/waifuisalie/RA3_1 into fix_atualizando_readme

breno-rossi committed:
7746288 Atualizando README

breno-rossi committed:
Atualizando README e Arquivos de Teste

README atualizado com extensiva documentação do projeto RA3 seguindo com os requisitos do trabalho. Tambem foram atualizado os arquivos de teste com 3 arqvuios de casos corretos e 3 casos incorretos sendo um refetente a erros nas 3 etapas do compilador desenvolvidos até o momento.

breno-rossi committed:
Atualizando README e  nome do sarquivos de teste

breno-rossi committed:
Atualizando README

fbleyruthes reviewed (approved):
Well done. README.md is complete.

breno-rossi committed:
Atualizando README

fbleyruthes reviewed (approved):
Now is well done. Unit tests include.

waifuisalie reviewed (approved):
looks good to me
